nrf: Add block device for VFS (v2)

dpgeorge opened the pull request:
This is a rebase of #5472 on latest master.

glennrub committed:
nrf: Facilitate use of freeze manifest.

Update the Makefile to handle FROZEN_MANIFEST, and the README with some
small samples on how to use freeze manifests.  And add BOARD_DIR to the
Makefile which can be referenced in boards/<board>/mpconfigboard.mk to
include a board specific manifest.

glennrub committed:
nrf/modules/nrf: Add new nrf module with flash block device.

This commit adds the "nrf" module for port specific modules and objects.
Included in it is the "Flash" object which exposes a block device
implementation to access internal SoC flash.

Thanks to @aykevl aka Ayke van Laethem for the initial implementation.

glennrub committed:
nrf/drivers: Add support for using flash block device with SoftDevice.

Update flash.c to also be compiled in when
MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE is enabled and SoftDevice is
present.

Update bluetooth/ble_drv.c to forward flash events to flash.c when
MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE is enabled.

glennrub committed:
nrf/mpconfigport.h: Expose nrf module when MICROPY_PY_NRF is set.

glennrub committed:
nrf/README: Update README.md to reflect internal file systems.

This documents parameters that can be passed to make to enable a specific
file system to included in the build.  Also, document the Makefile override
parameter "FS_SIZE" that can be used to tune the size of the flash region
to use as internal flash file system.

glennrub committed:
nrf/mpconfigport.h: Tune FAT FS configuration.

Disable MICROPY_FATFS_MULTI_PARTITION configuration because there is no
partition table in the flash for FATFS to read.

Also, set MICROPY_FATFS_MAX_SS to the size of a flash page.  For nrf51 the
value 1024 is set.  For nrf52/nrf91 the value 4096 is set.

glennrub committed:
nrf/Makefile: Add _fs_size linker script override from make.

Add posibility to override linker script "_fs_size" from make by adding the
FS_SIZE parameter.  The syntax of value is linker script syntax.  For
example, the value of 131072 bytes can be written as 128K like this:
FS_SIZE=128K.

If not set, default value for "_fs_size" from linker script will be used.

glennrub committed:
nrf/modules/uos: Allow a board to configure MICROPY_VFS_FAT/LFS1/LFS2.

glennrub committed:
nrf/mpconfigport.h: Enable MICROPY_PY_IO_FILEIO when an FS is enabled.

Enable MICROPY_PY_IO_FILEIO if MICROPY_VFS_FAT, MICROPY_VFS_LFS1
or MICROPY_VFS2 has been enabled.

glennrub committed:
nrf/qstrdefsport.h: Add entries for in-built FS mount points.

glennrub committed:
nrf/main: Add auto mount and auto format hook for internal flash FS.

glennrub committed:
nrf/boards: Enable needed features for FAT/LFS1/LFS2.

Enable the following features for all boards except
nrf51 boards with SoftDevice present:

- MICROPY_VFS
- MICROPY_PY_NRF
- MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE

glennrub committed:
nrf/boards: Set FROZEN_MANIFEST blank when SD present on nrf51 targets.

glennrub committed:
nrf/Makefile: Set default manifest file for all targets.

Set the default manifest to "modules/manifest.py".  This includes files
from the folder "modules/scripts".  The manifest default value is overriden
by all nrf51 boards that have SoftDevice present (SD=s110) to save flash.

Also add "modules/manifest.py" which is set to freeze
"modules/scripts/_mkfs.py".

glennrub committed:
nrf/boards: Enable MICROPY_VFS_LFS2 for all target boards.

Enable LittleFS v2 for all targets, except nrf51 targets when SoftDevice is
present.

glennrub committed:
nrf/modules/scripts: Add file system formatting script.

Add a helper script _mkfs.py which automatically formats the file system if
nrf.Flash() is located and a VFS file system has been included in the
compilation.

The precedence is: first LFS1, LFS2 then FAT.

glennrub committed:
tools/codeformat.py: Include ports/nrf/modules/nrf in code formatting.

glennrub committed:
tools/ci.sh: Add mpy-cross build to nrf port.

glennrub committed:
nrf/mphalport: Add dummy function for mp_hal_time_ns().

extmod/vfs_lfs.c needs to resolve `mp_hal_time_ns()` in order to calculate
a timestamp from 1970 epoch.  A wall clock is not available in the nrf
port, hence the function is implemented to resolve compilation linkage
error.  The function always return 0.

glennrub committed:
nrf/boards/common.ld: Calculate unused flash region.

Calculate the unused flash area on the target device.  The values will be
exposed by _unused_flash_start and _unused_flash_length.  The start address
and the length are not aligned to either word or pages.

mzdaniel commented:
Thank you for looking into this @dpgeorge. 

Compilation works straight now.  Here are some of the issues I found:

1) extmod/vfs_basic.py test is failing (regresion since #5472)

./run-tests.py --clean-failures
./run-tests.py --target nrf extmod/vfs_basic.py
./run-tests.py --print-failures| less

 stat /x OSError
 1 mount False False
 ['test_mnt']
-('test_mnt', 16384, 0)
-StopIteration
-StopIteration
...
+Traceback (most recent call last):
+  File "<stdin>", in <module>
+AttributeError: 'module' object has no attribute 'ilistdir'
+CRASH
\ No newline at end of file

FAILURE /data/micropython/repo/micropython/tests/results/extmod_vfs_basic.py


BTW may I ask for a review of https://github.com/micropython/micropython/pull/6395 ?  tail on the previous output would look as follow

-5 chdir /subdir
-OSError
-/
+Traceback (most recent call last):
+  File "<stdin>", line 90, in <module>
+AttributeError: 'module' object has no attribute 'ilistdir'
+CRASH
\ No newline at end of file

FAILURE /data/micropython/repo/micropython/tests/results/extmod_vfs_basic.py


2) The following code seems to freeze micropython. 

echo "import os; print(os.listdir()); os.mkdir('lib')" > mkdir_test
mpremote run mkdir_test 
[]
(... and at this point micropython freezes)

dpgeorge committed:
nrf/modules/uos: Add ilistdir to uos module.

This was missed in the initial implementation of the uos module.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Thanks @mzdaniel for testing.

I have added `uos.ilistdir` to fix point 1.

Point 2 was due to the flash size ioctl reporting bytes rather than blocks, and so littlefs thought the flash was much larger than it really is and so it attempted to read/write out-of-bounds flash.  This is now fixed.

I also tidied up a few other minor things.  So this should now be ready to go in.

mzdaniel commented:
Sound great! And thank you for your changes Damien.
I'll propose in a new PR changing the FLASH_PAGESIZE default and increasing fs size wherever appropriate.
